source/MaterialXCore/Element.cpp

```diff
@@ -104,6 +104,24 @@ string Element::getNamePath(ConstElementPtr relativeTo) const
     return res;
 }
 
+ElementPtr Element::getElementByPath(const std::string& path)
+{
+    std::vector<std::string> elementNames = splitString(path, NAME_PATH_SEPARATOR);
+    ElementPtr currentElement = getSelf();
+    if (elementNames.size() == 0)
+    {
+        return currentElement;
+    }
+    for (std::string elementName : elementNames)
+    {
+        if (!(currentElement = currentElement->getChild(elementName)))
+        {
+            return currentElement;
```

Comment:
If a path is given that only partially matches the hierarchy this will return the element at that sub-path right, and ignore the rest of the path? Would it not be better to require an exact match to the leaf element in the given path, and return nullptr if it's not found?

Reply:
Sorry my bad, currentElement is null in that case. But instead of returning currentElement you could explicitly return ElementPtr() to make the code more readable :)

Reply:
I don't think that's what would happen. If currentElement->getChild(elementName) fails, then this returns ElementPtr().

Reply:
Yep, that's right (as I wrote in my correction above), but I would still return ElementPtr() to make that more clear.

Here's a version with my suggested changes:

ElementPtr Element::getChildByNamePath(const std::string& path)
{
    const std::vector<std::string> elementNames = splitString(path, NAME_PATH_SEPARATOR);
    if (elementNames.empty())
    {
        return ElementPtr();
    }
    ElementPtr currentElement = getSelf();
    for (const std::string& elementName : elementNames)
    {
        if (!(currentElement = currentElement->getChild(elementName)))
        {
            return ElementPtr();
        }
    }
    return currentElement;
}

Note that empty string is now returning null instead of self, which feels more correct if it's a child element you're requesting.
Just my 2 cents.

Reply:
I'm fine with being more explicit and returning ElementPtr() instead of currentElement when the condition fails. I can see how people might miss that that's what it is doing.

For my purposes, I really want to get a Document when a user specifies document->getElementByPath(""); I agree it doesn't match the getChildByNamePath name very well though, but I didn't originally call it that. :) It makes sense that this would work though as document->getNamePath() would return "" and I'm just doing the opposite with document->getElementByPath("")

Reply:
Seems reasonable, I'll make the above suggestions soon.